docs: data studio updates (#72641)
metabase/metabase: docs/data-studio/dependencies/replace-data-sources.md

---
title: Replace data sources

> summary: Replace a data source across all questions, models, and other content that use it—in one operation.

Reviewer: ```suggestion
summary: Admins can swap out a table, model, or question and replace it with a different one across your entire Metabase.
```

---

# Replace data sources

{% include plans-blockquote.html feature="Replacing data sources" %}

> Admins can swap out a table, model, or saved question used across your Metabase and replace it with a different one across your entire Metabase. This is useful when you want to point existing content at a new or updated table, or when you're migrating from one data source to another.

Reviewer: ```suggestion
Admins can swap out a table, model, or question and replace it with a different one across your entire Metabase.

For example, let's say you created a transform that created a cleaned up version of your `orders` table, called `orders_clean`. Instead of rewriting all your questions that query the `orders` table, Admins can swap in the `order_clean` table everywhere the old `orders` table is referenced.
```

Think we should follow with a concrete example to give people a more concrete example. The original construction was kinda just restating data-source-swapping at a high level.

## How replacing data sources works

> - Only admins can replace data sources.

Reviewer: ```suggestion
```
We mention admins elsewhere.

What about Data Analyst group?

Author: People don't read.Also yes it's only admins,  not data analysts.

> - Metabase uses the [dependency graph](./graph.md) to identify all content that depends on the original data source. "Data source" here means a table, model, or saved question - anything that's used as, well, a data source for a query.

Reviewer: ```suggestion
- Metabase uses the [dependency graph](./graph.md) to identify all content that depends on the original data source. "Data source" here could be a table, model, or question queried by another entity.
```

- Once an admin [selects a replacement source](#find-and-replace-data-sources), Metabase [checks compatibility](#compatibility-requirements) of the original data source with the replacement data source.

- If data sources are compatible, Metabase walks through the entire dependency graph for the original data source and changes all queries to use the new data source instead. Metabase only swaps data sources where they're used specifically as _data sources_ for queries - and not, for example, lists of values or custom destinations, see [What gets replaced](#what-gets-replaced).

- The original data source is _not deleted_ after the replacement is finished.

## Compatibility requirements

You can replace a table, model, or saved question with any other table, model, or saved question, as long as the two sources satisfy the following requirements:

### General requirements

> - **Original and new data sources must be in the same database**.
> 
>   For questions/models, this means that the underlying tables must be in the same database.

Reviewer: ```suggestion
- **Original and new data sources must query the same database**. Questions and models can only be swapped if they query tables in the same database.
```

Also, wonder if we can inline these list items like in the suggested edit, just to save vertical scrolling.

Author: I tried that but paragraphs become too long and visually overwhelming. I immediately think "oh this is going to be a drag to read, i'll just skip it for now" once i see paragraphs like that

- **Column types must match**.

  This means that, for example, if the column `user_id` is an integer in the old data source, it also must be an integer (and not a string or UUID) in the new data source.

- **If the original column is a Metabase _foreign_ key, the replacement column must be too.**

  For example, if `old_table` has `User ID` column with Metabase semantic type "Foreign key" to the `ID` column of `users` table, then the replacement source `new_table` also must have a `User ID` column with Metabase foreign key to `ID` from `users`.

  Only foreign keys defined in Metabase need to match. Database-defined foreign keys aren't checked.

- **If the original column is a Metabase _entity_ key, the replacement column must be too.**

  For example, if `old_table` has `ID` column with Metabase semantic type "Entity key", then the replacement source `new_table` also must have an `ID` column with Metabase semantic type "Entity key".

- If replacing a table, that **original table can't be a target of a Metabase's [semantic foreign keys](../../data-modeling/semantic-types.md)** from other tables.

>   Foreign keys can still exist on the database level, but not on the level of Metabase semantic types. This requirement is necessary to avoid issues with implicit joins when swapping tables: for example, there can

Reviewer: Cuts off here.

  The old table can still have foreign keys _to_ other tables, it just can't be the _target_ of a foreign key.

> ## Find and replace data sources

Reviewer: This should be the first section.

The above stuff only needs to be read if they run into a problem.

{% include plans-blockquote.html feature="Dependency graph" %}

Only admins can replace data sources.

To find and replace all entities that depend on a data source:

1. Go to **Data Studio** by clicking the **grid icon** in the top right and selecting **Data Studio**;
2. In Data Studio, click **Dependency graph** in the sidebar and search for the table, model, or question you want to replace.
3. Click the card for the data source and in the info panel on the right, click the **Find and replace** icon.
   ![Find and replace](../images/find-and-replace.png)

> 4. In the left sidebar, select replacement data source.

Reviewer: Might want to link to the permissions section here somewhere.

5. If Metabase can't go ahead with replacement, you'll see error messages next to data sources, and column compatibility errors in the "Column comparison table".

   ![Compatibility requirements](../images/compatibility-requirements.png)

>    Check out [Compatibility requirements](#compatibility-requirements).

Reviewer: ```suggestion
5. If Metabase can't go ahead with replacement, you'll see error messages next to data sources, and [compatibility requirements](#compatibility-requirements) errors in the "Column comparison table".

   ![Compatibility requirements](../images/compatibility-requirements.png)
```

6. If the sources are compatible, you can check out the list of items that will be changed:

>    ![Preview changes](../images/check-repalce-changes.png)

Reviewer: Typo

> 7. Once you verified the changes, click U&&

Reviewer: > U&&!

Author: sigh, i thought i did a typo scan

### Running the replacement

> Replacement can take a long time (potentially hours) on instances with thousands of questions.

Reviewer: ```suggestion
If your Metabase has thousands of questions, replacement can take a long time (potentially hours).
```

Metabase runs the replacement asynchronously in the background. A status indicator appears at the bottom of the screen showing progress and completion. When complete, all affected items are updated to use the new data source.

If the replacement fails partway through, some items may already be updated. Check the status indicator for details.

You can cancel a replacement while it's running. Items that were already updated before you canceled won't be reverted.

> Only one replacement can run at any time.

Reviewer: Yikes. Should they just try again? Or to revert, try swapping in the opposite direction? What's our recommendation here?

> ## What gets replaced
> 
> Items are replaced when (and only when) they are used as a query source. This includes:

Reviewer: ```suggestion
## Only query _sources_ are replaced

Query sources include:
```

- Data sources (including in joins) in questions, models, and metrics built with the query builder;
- Data sources for measures and segments;
- Tables in `FROM` statements in SQL queries;
- [Questions and models referenced from SQL questions](../../questions/native-editor/referencing-saved-questions-in-queries.md);

> - Field filters;

Reviewer: ```suggestion
- Data sources (including in joins) in questions, models, and metrics built with the query builder
- Data sources for measures and segments
- Tables in `FROM` statements in SQL queries
- [Questions and models referenced from SQL questions](../../questions/native-editor/referencing-saved-questions-in-queries.md)
- Field filters
```

> Metabase **will not replace** tables/models/questions in:

Reviewer: ```suggestion
Replacements exclude:
```

Author: I use active voice for a reason - to emphasize the fact that metabase will actively skip these things (even if it finds dependencies on them)

- Snippets

> - Question- or model-based sources for dropdown filter values in SQL questions and dashboards

Reviewer: ```suggestion
- Questions or models that power dropdown filter values in SQL questions and dashboards
```
Is this right?

- [Links](../../dashboards/introduction.md#link-cards) and [custom destinations](../../dashboards/interactive.md#custom-destinations) on dashboards;
- Targets for [actions](../../actions/introduction.md).

> Metabase will **not allow** you to replace data sources that are used as:

Reviewer: ```suggestion
You can't replace data sources that are used as:
```

Author: This is written like that for a reason - to distinguish "replacement is ran, but some things are skipped" from "replacement is not ran at all"

- Row or column security policies
- Targets of foreign keys

> ## Permissions and data source replacement

Reviewer: ```suggestion
## Swapped-in data source should have comparable permissions
```

Or

> Make sure the swapped data source has comparable permissions

Just to put the point in the heading.

Author: That heading will be too long to scan

> Only admins can replace data sources in bulk.

Reviewer: ```suggestion
```

I'd move this to the end of the section. It's relevant, but stated elsewhere, so somewhat buries the lede of this section.

> Metabase **will not check** whether the data sources have comparable permissions. You need to make sure that new data source has appropriate permissions _before_ initiating the replacement to avoid blocking people from seeing questions, or exposing data they're not supposed to see.

Reviewer: ```suggestion
Metabase **won't check** whether swapped data sources have comparable permissions. Admins need to make sure that new data source has appropriate permissions _before_ initiating the replacement, otherwise you could block people from viewing questions, or expose data they're not supposed to see.
```

> For example, `old_table` might be configured to have view access for all users, but `new_table` might be blocked for everyone. If you replace `old_table` with `new_table`, then people will not be able to see any data in questions that used the `old_table` before replacement, because those questions will now use the blocked `new_table`.

Reviewer: Good example, but I would rephrase to talk explicitly about groups.
```suggestion
For example, the Customer group might have view data access to `old_table`, but have blocked (no) access to `new_table`. If you replace `old_table` with `new_table`, then people in the Customers group won't be able to see results in questions that used to query the `old_table` and now query the blocked `new_table`.
```

> Review both data and collection permissions for old and new sources before starting replacement.

Reviewer: Maybe a note to explain why you'd need to vet collection permissions. (Assuming it's in situations where people swap one question for another).

> ## Avoid using data source replacement with remote sync enabled

Reviewer: ```suggestion
## Avoid using data source replacement with remote sync
```
Good heading.

But should this be it's own section? Why not not just a limitation?

Author: It's not a good heading because it's too long to scan but also the section is kind of meh. The only reason it's very important - stuff might get messed up if you use replacement with remote sync

Because old data sources can be used both in synced and unsynced collections, we do not recommend swapping data sources in bulk when remote sync is enabled.

## Limitations

- The source and target must be on the same database. Cross-database replacement isn't supported.
- Original and new data source have to satisfy [compatibility requirements](#compatibility-requirements).

> - All uses of the original data sources are swapped for a new data source. You can't pick and choose the content to replace data sources in.

Reviewer: ```suggestion
- All uses of the original data source is swapped for the new data source. You can't pick and choose which content to update.
```

- **Column types must match**.

  This means that, for example, if the column `user_id` is an integer in the old data source, it also must be an integer (and not a string or UUID) in the new data source.

- **If the original column is a Metabase _foreign_ key, the replacement column must be too.**

  For example, if `old_table` has `User ID` column with Metabase semantic type "Foreign key" to the `ID` column of `users` table, then the replacement source `new_table` also must have a `User ID` column with Metabase foreign key to `ID` from `users`.

  Only foreign keys defined in Metabase need to match. Database-defined foreign keys aren't checked.

- **If the original column is a Metabase _entity_ key, the replacement column must be too.**

  For example, if `old_table` has `ID` column with Metabase semantic type "Entity key", then the replacement source `new_table` also must have an `ID` column with Metabase semantic type "Entity key".

- If replacing a table, that **original table can't be a target of a Metabase's [semantic foreign keys](../../data-modeling/semantic-types.md)** from other tables.

>   Foreign keys can still exist on the database level, but not on the level of Metabase semantic types. This requirement is necessary to avoid issues with implicit joins when swapping tables: for example, there can

Author: ```suggestion
  Foreign keys can still exist on the database level, but not on the level of Metabase semantic types. This requirement is necessary to avoid issues with implicit joins when swapping tables.
```
